### PR TITLE
Fixes #109 After 'oc delete rc ...', remote JVM on linux was still ru…

### LINKED ISSUE #109: After 'oc delete rc ...', remote JVM on linux was still running.
We likely need to do a graceful shutdown of the SSH session so that the remote ssh server notices that the session has been killed so that the process can receive the HUP signal.

### PR DESCRIPTION
**Pull request by chirino**
…nning.

We not do a close on the ssh session before exiting.

### CODE REVIEW

File: ssh/ssh.go [hunk 1]
```diff
@@ -90,9 +92,24 @@ func RemoteSSHCommand(user string, privateKey string, host string, port string,
 		}
 	}
 
-	log.Info("Running command %s", cmd)
-	err = session.Run(cmd)
-	if err != nil {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
```
Comment (jimmidyson) [inline ssh/ssh.go:96]: This should only be SIGINT & SIGTERM.

File: ssh/ssh.go [hunk 2]
```diff
@@ -90,9 +92,24 @@ func RemoteSSHCommand(user string, privateKey string, host string, port string,
...
+	// Run in a go routine so that we can wait for it to finish or a OS signal to get raised.
+	go func() {
+		log.Info("Running command %s", cmd)
+		runnerr = session.Run(cmd)
+		signals <- syscall.SIGCHLD
+	}()
+
+	signal := <- signals
+	if runnerr != nil {
+		if signal!= syscall.SIGCHLD {
```
Comment (jimmidyson) [inline ssh/ssh.go]: This file needs `go fmt`ing.

File: ssh/ssh.go [hunk 3]
```diff
@@ -90,9 +92,24 @@ func RemoteSSHCommand(user string, privateKey string, host string, port string,
...
-	if err != nil {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
+
+	var runnerr error = nil;
+	// Run in a go routine so that we can wait for it to finish or a OS signal to get raised.
+	go func() {
+		log.Info("Running command %s", cmd)
+		runnerr = session.Run(cmd)
+		signals <- syscall.SIGCHLD
```
Comment (jimmidyson) [inline ssh/ssh.go]: Can we do this with a sync.WaitGroup instead?

File: ssh/ssh.go (hunk 2 above)
Comment (jimmidyson) [inline ssh/ssh.go]: SIGCHLD is more for forked child processes, not goroutines.

File: ssh/ssh.go [hunk 4]
```diff
@@ -90,9 +92,24 @@ func RemoteSSHCommand(user string, privateKey string, host string, port string,
...
+	}()
+
+	signal := <- signals
+	if runnerr != nil {
+		if signal!= syscall.SIGCHLD {
+			log.Info("Rceived OS Signal %v. Closing SSH session.", signal)
+		}
+	}
+	if runnerr != nil {
```
Comment (jimmidyson) [inline ssh/ssh.go]: This duplicates the previous condition - just add logic to previous block.

**Comment by jimmidyson** [comment]
Wouldn't this be easier to do the ssh logic in the main & the signal catching in the goroutine, closing the ssh session when catching the expected signals?

File: ssh/ssh.go (hunk 2 above)
Comment (chirino) [inline ssh/ssh.go]: so when the run command finishes I want the guy which is waiting for a signal to unblock.  I just picked a signal that was not being used and sent that.

**Comment by chirino** [comment]
I'll try switching it around.

File: ssh/ssh.go (hunk 1 above)
Comment (chirino) [inline ssh/ssh.go:96]: All those are valid signals that typically indicate the the user wants the process to shutdown.

Comment (jimmidyson) [inline ssh/ssh.go:96]: SIGHUP signifies the controlling terminal is closed & is normally used by daemons to signify reloading config, not stopping.

Happy to handle syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT though. Note that [as documented](https://golang.org/pkg/os/#Signal) only SIGINT & SIGKILL are guaranteed to exist on all platforms.

**Comment by jimmidyson** [comment]
ok to test

**Pull request merged**